mesonpy/_editable.py

```diff
@@ -2,183 +2,151 @@
-    """Custom loader that whose purpose is to detect when the import system is
-    trying to load our modules, and trigger a rebuild. After triggering a
-    rebuild, we return None in find_spec, letting the normal finders pick up the
-    modules.
-    """
-
-    def __init__(
-        self,
-        project_name: str,
-        hook_name: str,
-        project_path: str,
-        build_path: str,
-        import_paths: List[str],
-        top_level_modules: List[str],
-        rebuild_commands: List[List[str]],
-        verbose: bool = False,
-    ) -> None:
-        self._project_name = project_name
-        self._hook_name = hook_name
-        self._project_path = project_path
-        self._build_path = build_path
-        self._import_paths = import_paths
-        self._top_level_modules = top_level_modules
-        self._rebuild_commands = rebuild_commands
+    NodeBase = dict
+
+
+MARKER = 'MESONPY_EDITABLE_SKIP'
+VERBOSE = 'MESONPY_EDITABLE_VERBOSE'
+
+LOADERS = [
+    (importlib.machinery.ExtensionFileLoader, tuple(importlib.machinery.EXTENSION_SUFFIXES)),
+    (importlib.machinery.SourceFileLoader, tuple(importlib.machinery.SOURCE_SUFFIXES)),
+    (importlib.machinery.SourcelessFileLoader, tuple(importlib.machinery.BYTECODE_SUFFIXES)),
+]
+
+
+class Node(NodeBase):
+    def __missing__(self, key: str) -> Node:
+        value = self[key] = Node()
+        return value
+
+    def __setitem__(self, key: Union[str, Tuple[str, ...]], value: Union[Node, str]) -> None:
+        n = self
+        if isinstance(key, tuple):
+            for k in key[:-1]:
+                n = n[k]  # type: ignore
+            key = key[-1]
+        dict.__setitem__(n, key, value)
+
+    def __getitem__(self, key: Union[str, Tuple[str, ...]]) -> Union[Node, str]:
+        n = self
+        if isinstance(key, tuple):
+            for k in key[:-1]:
+                n = n[k]  # type: ignore
+            key = key[-1]
+        return dict.__getitem__(n, key)
+
+    def get(self, key: Union[str, Tuple[str, ...]]) -> Optional[Union[Node, str]]:  # type: ignore[override]
+        n = self
+        if isinstance(key, tuple):
+            for k in key[:-1]:
+                n = dict.get(n, k)  # type: ignore
+                if n is None:
+                    return None
+            key = key[-1]
+        return dict.get(n, key)
+
+
+def walk(root: str, path: str = '') -> Iterator[pathlib.Path]:
+    with os.scandir(os.path.join(root, path)) as entries:
+        for entry in entries:
+            if entry.is_dir():
+                yield from walk(root, os.path.join(path, entry.name))
+            else:
+                yield pathlib.Path(path, entry.name)
+
+
+def collect(install_plan: Dict[str, Dict[str, Any]]) -> Node:
+    tree = Node()
+    for key, data in install_plan.items():
+        for src, target in data.items():
+            path = pathlib.Path(target['destination'])
+            if path.parts[0] in {'{py_platlib}', '{py_purelib}'}:
```

Comment:
What's wrong with a set?

Reply:
The general wisdom is to always prefer a set, surely... set membership is faster.


IIRC:

For small constants, tuples are very close to equal performance, depending on your version of python... sets are more expensive to create which can tilt the balance in favor of a tuple depending on your inputs, but both can be constructed in the compiler instead of the interpreter so it's a moot point here.